src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractTransactionService.cs

```diff
@@ -199,36 +200,101 @@ public BuildContractTransactionResult BuildTx(BuildContractTransactionRequest re
+        /// <param name="param">The string to parse.</param>
+        /// <param name="ndx">The position in the string following the '[' character.</param>
+        /// <returns>The elements as a string array.</returns>
+        /// <remarks>
+        /// Single quotes are removed from elements and double single quotes are converted to literal single quotes.
+        /// </remarks>
+        public static string[] ParseArray(string param, ref int ndx)
         {
-            const int signatureLength = 65;
-            const int minHeaderByte = 27;
-            const int maxHeaderByte = 34;
+            var elements = new List<string>();
+            int elementStart = ndx;
+            bool quoted = false;
 
+            for (; ndx < param.Length; ndx++)
+            {
+                if (param[ndx] == '\'')
+                {
+                    quoted = !quoted;
+                }
+                else if (!quoted && (param[ndx] == ',' || param[ndx] == ']'))
+                {
+                    // Check for element termination characters. 
+                    string element = param.Substring(elementStart, ndx - elementStart).Trim().Replace("''", "\x0").Replace("'", "").Replace("\x0", "'");
+                    elements.Add(element);
+                    if (param[ndx] == ']')
+                        return elements.ToArray();
+                    elementStart = ndx + 1;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Replaces parameters of the form "N1[element1,element2,element3,...]" with byte arrays of the form "N2#byte-array-in-hex",
+        /// where N1 and N2 are the integer values of the <see cref="MethodParameterDataType"/>, with N2 being 10 (byte array).
+        /// </summary>
+        /// <param name="network">The network.</param>
+        /// <param name="parameters">A list of parameters that may contain arrays.</param>
+        /// <returns>The input parameters with arrays replaced with byte arrays.</returns>
+        /// <remarks>The encoded byte arrays are suitable for deserializing as arrays of the specified type (N1 in this example)
+        /// by using <see cref="ContractPrimitiveSerializer.Deserialize{T}(byte[])"></see>.</remarks>
+        private static string[] ReplaceArraysWithByteArrays(Network network, string[] parameters)
+        {
             if (parameters == null)
                 return null;
 
-            // Replace SIG# with any included signatures.
-            string encodedSigs = null;
+            var cpSerializer = new ContractPrimitiveSerializer(network);
+            var mpSerializer = new MethodParameterStringSerializer(network);
+
+            // Replace arrays with byte arrays.
             for (int i = 0; i < parameters.Length; i++)
             {
-                if (parameters[i].ToUpper() == "SIG#")
+                string param = parameters[i];
+
+                // Parse the type.
+                int type = 0;
```

Comment:
Use `MethodParameterDataType` instead of `int`

Reply:
Constant added.